src/sagas/wallet.js

```diff
@@ -386,24 +387,71 @@ export function* handleTx(action) {
   }
 
   // find tokens affected by the transaction
-  const balances = yield call(wallet.getTxBalance.bind(wallet), tx);
   const stateTokens = yield select((state) => state.tokens);
   const registeredTokens = stateTokens.map((token) => token.uid);
 
-  // We should download the **balance** for every token involved in the transaction
-  // and history for hathor and DEFAULT_TOKEN
-  for (const [tokenUid] of Object.entries(balances)) {
-    if (registeredTokens.indexOf(tokenUid) === -1) {
-      continue;
+  // To be able to only download balances for tokens belonging to this wallet, we
+  // need a list of tokens and addresses involved in the transaction from both the
+  // inputs and outputs.
+  const { inputs, outputs } = tx;
+  const data = [...inputs, ...outputs];
+
+  // tokenAddressesMap should be an object { [tokenUid]: Set(addresses) }
+  // txAddresses should be a Set with all addresses involved in the tx
+  const [tokenAddressesMap, txAddresses] = data.reduce(
+    (acc, io) => {
+      if (!io.decoded) {
+        return acc;
+      }
+
+      const { token, decoded: { address } } = io;
```

Comment:
```suggestion
      const { token, decoded: { address } } = io;
      
      if (!address) {
        return acc;
      }
```

If the address is not defined, should we ignore the `input/output`?
I think this can happen with data outputs.

Reply:
Fixed! Thanks

Reply:
I think we should. Fixed in c36989f17dcee6be4c254f94a8e6ec5844d97fe1